### ISSUE TITLE
Extremely poor build performance on vertically scaled hardware

### ISSUE THREAD

**Issue by mcumings** [gradle/gradle#16211]
### Expected Behavior

Build should perform better than the same build on a MacBook Pro when run on a beast of a vertically scaled piece of hardware.

### Current Behavior

Build takes significantly longer, apparently due to issues with resource locking.

Dumping threads in the middle of a slow build ( 
[gradlestack.txt](https://github.com/gradle/gradle/files/6003676/gradlestack.txt)
 ) led me to believe this was related to #6696 but Gary thought it might differ due to the fact that my build is proceeding slowly whereas the the other issue describes a hang.  In the thread dump it appears as if all 78 worker threads are in: `org.gradle.internal.resources.DefaultResourceLockCoordinationService.withStateLock`

### Context

I just received a new piece of hardware to test the upper scaling limits of our build.  The machine has 512GB of RAM, 80 CPU threads, and NVMe storage.  When attempting to run a build on this machine the build ran significantly slower than on my corpware-laden MacBook Pro.

The project is a large Android application.

### Steps to Reproduce 

This would likely repro on any large project with a significantly scaled piece of hardware.  I don't think this would be project-specific from the symptoms I'm observing.

### Your Environment

I've also run gradle-profiler with the async profiler enabled.  Results also attached. [gradle-profiler.tar.gz](https://github.com/gradle/gradle/files/6003649/gradle-profiler.tar.gz)

I've captured and attached a build scan of the final profiled build (
[vo4ysv2zz43ba.scandump.gz](https://github.com/gradle/gradle/files/6003657/vo4ysv2zz43ba.scandump.gz) )
.  This will describe our environment in depth.  Note that this issue even occurred without enabling `--parallel` builds.  The scan dump shows significant delays between tasks in the timeline view.

**Comment by ghale** [comment]
I'm fairly certain that this is related to the number of workers combined with the size of the task graph.  Basically, all of those workers are crawling through the task graph looking for tasks that are ready to run.  Because of the way this is architected, they're holding a lock around a critical section that prevents the tasks from signaling that another task can now run.  The amount of time they spend in that section is related to the number of tasks in the graph.  More workers means more contention, more tasks means they hold it longer.

Although I'm sure the problem is still there when run with parallel enabled, I suspect that the problem isn't as "bad".  Unlike the not-parallel case, workers would at least find tasks in other projects that can be executed while all this contention is taking place.  Does that match what you're seeing (i.e. "better", but still impacted, performance with parallel enabled)?

I think fixing this will require some fundamental changes to how we walk the task graph.

**Comment by mcumings** [comment]
Hey Gary, thanks for taking a gander!

Unfortunately, when enabling --parallel performance completely fell off a cliff (5h44m51s to complete the 15 minute build).

I ended up raising a gradle enterprise support ticket for this to get official support on it.  Hopefully someone can come up with some sort of resolution... 😄

**Comment by ghale** [comment]
One experiment to try would be to set `--max-workers` to something similar to what you have on the macbook and try it on your "beast" machine.  This would tell us whether the performance degradation is due to the extra workers or due to something else.

**Comment by ghale** [comment]
Is it possible in the `--parallel` build you were bumping up against heap space (build scan would reveal this).  Perhaps with more stuff running in parallel, you started contending for available heap?

**Comment by big-guy** [comment]
@ghale we have that experiment already:

8m39s (beast) vs 11m59s (mac)

I also have the build scans I can share with you internally.

**Comment by mcumings** [comment]
This issue was resolved by Gradle 6.8.3.

Closing.
